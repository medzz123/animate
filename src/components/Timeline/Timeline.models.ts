export interface TimelineProps {
  handleReset: () => void;
}
