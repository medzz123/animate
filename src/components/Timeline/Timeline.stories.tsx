// import { Meta, Story } from '@storybook/react/types-6-0';
// import React from 'react';

// import Timeline from './Timeline';

// export default {
//   title: 'Components/Timeline',
//   component: Timeline,
// } as Meta;

// const Template: Story = (storyArguments) => <Timeline {...storyArguments} />;

// export const TimelineStory = Template.bind({});
