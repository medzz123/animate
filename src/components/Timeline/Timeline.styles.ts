import styled, { css } from 'styled-components';

import { buttonReset } from '../../theme/resets';
import { tokens } from '../../theme/tokens';

export const DeleteButton = styled.button`
  margin-left: ${tokens.sizes[8]};
  width: ${tokens.sizes[20]};
  height: ${tokens.sizes[20]};
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: ${tokens.sizes.half};
  background-color: ${(p) => p.theme.background};
  color: ${(p) => p.theme.danger};
  box-shadow: ${(p) => p.theme.shadow};

  svg {
    width: ${tokens.sizes[12]};
    height: ${tokens.sizes[12]};
  }
`;

export const PlayStateButtons = styled.button.attrs({ type: 'button' })`
  color: ${(p) => p.theme.main};

  svg {
    width: ${tokens.sizes[24]};
    height: ${tokens.sizes[24]};
  }
`;

export const PillInputContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
`;

export const PillInput = styled.input`
  width: ${tokens.sizes[70]};
  height: ${tokens.sizes[28]};
  border-radius: ${tokens.sizes[16]};
  background-color: ${(p) => p.theme.backgroundContrast};
  box-shadow: none;
  border: none;
  outline: none;

  color: ${(p) => p.theme.main};
  font-weight: bold;
  padding: ${tokens.sizes[8]};
`;

export const PillButton = styled.button.attrs({ type: 'button' })<{
  move?: boolean;
}>`
  border-radius: ${tokens.sizes.half};
  height: ${tokens.sizes[28]};
  width: ${tokens.sizes[28]};
  display: flex;
  justify-content: center;
  align-items: center;

  ${(p) =>
    p.move &&
    css`
      transform: translateX(-${tokens.sizes.full});
    `}

  background-color: ${(p) => p.theme.main};

  color: ${(p) => p.theme.background};

  svg {
    width: ${tokens.sizes[16]};
    height: ${tokens.sizes[16]};
  }
`;

export const TimelineContainer = styled.div`
  width: ${tokens.sizes.full};
  background-color: ${(p) => p.theme.backgroundContrast};

  height: ${tokens.sizes.full};
  overflow-y: scroll;

  -ms-overflow-style: none; /* IE and Edge */
  scrollbar-width: none; /* Firefox */

  box-shadow: ${(p) => p.theme.shadow};
  border-radius: ${tokens.sizes[20]};

  span {
    color: ${(p) => p.theme.paragraph};
  }

  &::-webkit-scrollbar {
    display: none;
  }
`;

export const ActionsContainer = styled.div`
  padding: ${tokens.sizes[24]} ${tokens.sizes[40]};
  display: flex;
  justify-content: space-between;
  align-items: center;

  svg {
    height: ${tokens.sizes[16]};
  }
`;

export const List = styled.ul`
  margin: 0;
  flex-direction: column;
  align-items: flex-start;
  width: ${tokens.sizes.full};
`;

export const Item = styled.li<{ selected?: boolean }>`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;

  transition: all ${tokens.transitions.fast} ease;

  padding: ${tokens.sizes[16]} ${tokens.sizes[40]};

  width: ${tokens.sizes.full};

  border-radius: ${tokens.sizes[20]};

  ${(p) =>
    p.selected &&
    css`
      background-color: ${(p) => p.theme.focus};
    `}
`;

export const Overflow = styled.span`
  font-size: ${tokens.sizes[14]};
  font-weight: bold;
  min-width: ${tokens.sizes[52]};
`;

export const TimelineText = styled.span`
  margin-right: ${tokens.sizes[30]};
  font-size: ${tokens.sizes[12]};
  color: ${(p) => p.theme.paragraph};
  min-width: ${tokens.sizes[70]};
`;

export const FramesContainer = styled.div`
  height: ${tokens.sizes[24]};
  width: ${tokens.sizes.full};
  position: relative;

  &::before {
    position: absolute;
    left: 0;
    top: ${tokens.sizes.half};
    content: '';
    width: ${tokens.sizes.full};
    height: ${tokens.sizes[1]};
    background-color: ${(p) => p.theme.main}${tokens.alpha[65]};
  }
`;

export const Frame = styled.button<{ position: string; selected: boolean }>`
  ${buttonReset}
  position: absolute;
  transition: all ${tokens.transitions.fast} ease;
  left: ${(p) => p.position}%;
  top: ${tokens.sizes.half};
  cursor: pointer;

  background-color: ${(p) => p.theme.main};
  width: ${tokens.sizes[7]};
  height: ${tokens.sizes[7]};
  border-radius: ${tokens.sizes.half};

  transform: translateY(-${tokens.sizes.half});

  ${(p) =>
    p.selected &&
    css`
      box-shadow: none;
      transform: translateY(-${tokens.sizes.half}) scale(2);
    `}

  &:focus,
  &:hover {
    transform: translateY(-${tokens.sizes.half}) scale(2);
  }
`;
