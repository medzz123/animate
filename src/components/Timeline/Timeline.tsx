import React, { FunctionComponent } from 'react';
import { VscClose } from 'react-icons/vsc';
import { useTheme } from 'styled-components';

import { useAnimationState } from '../../state/Animation/animation';
import Box from '../Box';
import {
  DeleteButton,
  Frame,
  FramesContainer,
  Item,
  List,
  Overflow,
  TimelineContainer,
} from './Timeline.styles';

const Timeline: FunctionComponent = () => {
  const {
    state,
    setElement,
    setStep,
    currentStep,
    deleteElement,
  } = useAnimationState();

  const theme = useTheme();

  return (
    <TimelineContainer data-testid="timeline">
      <List>
        {Object.keys(state.elements).map((key) => (
          <Item
            key={key}
            selected={key === state.element}
            onClick={() => {
              setElement(key);
            }}
          >
            <Overflow>{key}</Overflow>
            {key !== state.element ? (
              <DeleteButton
                type="button"
                onClick={(event) => {
                  event.stopPropagation();
                  deleteElement(key);
                }}
              >
                <VscClose size="20px" color={theme.paragraph} />
              </DeleteButton>
            ) : (
              <Box minWidth={48} height={24} />
            )}
            <FramesContainer>
              {Object.keys(state.elements[key].steps).map((frame) => (
                <Frame
                  selected={
                    key === state.element && Number(frame) === currentStep
                  }
                  key={`frame@${frame}`}
                  position={frame}
                  type="button"
                  onClick={() => {
                    setStep(Number(frame));
                  }}
                />
              ))}
            </FramesContainer>
          </Item>
        ))}
      </List>
    </TimelineContainer>
  );
};

export default Timeline;
