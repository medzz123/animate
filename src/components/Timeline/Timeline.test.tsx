import React from 'react';

import { render } from '../../../__tests__/testUtils';
import Timeline from './Timeline';

test('Timeline Test', () => {
  const { container } = render(<Timeline>Hello</Timeline>);

  expect(container).toMatchSnapshot();
});
