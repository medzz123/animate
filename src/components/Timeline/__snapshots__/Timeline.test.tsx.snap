// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline Test 1`] = `
<div>
  <div
    class="sc-dkPtyc bepIQA"
    data-testid="timeline"
  >
    <ul
      class="sc-hKwCoD dLNPKa"
    >
      <li
        class="sc-eCImvq hiimFq"
      >
        <span
          class="sc-jRQAMF ibcQbN"
        >
          square
        </span>
        <div
          class="sc-bdvvaa kgncym"
          data-testid="box"
        />
        <div
          class="sc-iCfLBT fYESJe"
        >
          <button
            class="sc-furvIG jBUuLZ"
            type="button"
          />
          <button
            class="sc-furvIG GMUcJ"
            type="button"
          />
          <button
            class="sc-furvIG kFlig"
            type="button"
          />
        </div>
      </li>
      <li
        class="sc-eCImvq jkLjNy"
      >
        <span
          class="sc-jRQAMF ibcQbN"
        >
          circle
        </span>
        <button
          class="sc-gsDJrp fnAnkT"
          type="button"
        >
          <svg
            color="#ffffff"
            fill="currentColor"
            height="20px"
            stroke="currentColor"
            stroke-width="0"
            style="color: rgb(255, 255, 255);"
            viewBox="0 0 16 16"
            width="20px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M8 8.707l3.646 3.647.708-.707L8.707 8l3.647-3.646-.707-.708L8 7.293 4.354 3.646l-.707.708L7.293 8l-3.646 3.646.707.708L8 8.707z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
        <div
          class="sc-iCfLBT fYESJe"
        >
          <button
            class="sc-furvIG jBUuLZ"
            type="button"
          />
          <button
            class="sc-furvIG hSgCnE"
            type="button"
          />
          <button
            class="sc-furvIG kFlig"
            type="button"
          />
        </div>
      </li>
    </ul>
  </div>
</div>
`;
